File: aemet/SergioDeIscar-aemet-java/SergioDeIscar-aemet-java/src/main/java/org/example/repositories/informe/InformeRepositoryMap.java
package org.example.repositories.informe;

import org.example.models.Informe;
import org.example.services.storage.informe.InformeStorageService;

import java.time.LocalDate;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class InformeRepositoryMap implements InformeRepository{
    private final InformeStorageService storageService;

    public InformeRepositoryMap(InformeStorageService storageService) {
        this.storageService = storageService;
    }
    private Map<LocalDate, Informe> informes = new LinkedHashMap<>();

    @Override
    public List<Informe> getAll() {
        upgrade();
        return informes.values().stream().toList();
    }

    @Override
    public Informe getById(LocalDate localDate) {
        return informes.get(localDate);
    }

    @Override
    public Informe save(Informe element, boolean storage) {
        informes.put(element.getDay(), element);
        if (storage) downgrade();
        return element;
    }

    @Override
    public void saveAll(List<Informe> elements, boolean storage) {
        elements.forEach(e -> save(e, storage));
    }

    @Override
    public Informe deleteById(LocalDate localDate) {
        return informes.remove(localDate);
    }

    @Override
    public Informe update(Informe element) {
        return updateById(element.getDay(), element);
    }

    @Override
    public Informe updateById(LocalDate localDate, Informe element) {
        if (getById(localDate) == null) return null;
        informes.put(localDate, element);
        downgrade();
        return element;
    }

    @Override
    public List<Informe> upgrade() {
        informes.clear();
        var load = storageService.loadAll();
        saveAll(load, false);
        return load;
    }

    @Override
    public List<Informe> downgrade() {
        return storageService.saveAll(informes.values().stream().toList());
    }
}

File: aemet/SergioDeIscar-aemet-kotlin/SergioDeIscar-aemet-kotlin/src/main/kotlin/repositories/informe/InformeRepositoryMap.kt
package repositories.informe

import models.Informe
import mu.KotlinLogging
import services.storage.informe.InformeStorageService
import java.time.LocalDate

private val logger = KotlinLogging.logger {  }

class InformeRepositoryMap(
    private val storageService: InformeStorageService
): InformeRepository {
    private val informes = mutableMapOf<LocalDate,Informe>()

    override fun getAll(): List<Informe> {
        logger.debug { "InformeRepositoryMap ->\tgetAll" }
        return informes.values.toList()
    }

    override fun getById(id: LocalDate): Informe? {
        logger.debug { "InformeRepositoryMap ->\tgetById" }
        return informes[id]
    }

    override fun save(element: Informe, storage: Boolean): Informe {
        logger.debug { "InformeRepositoryMap ->\tsave" }
        informes[element.day] = element
        if (storage) downgrade()
        return element
    }

    override fun saveAll(elements: List<Informe>, storage: Boolean) {
        logger.debug { "InformeRepositoryMap ->\tsaveAll" }
        elements.forEach { save(it, storage) }
    }

    override fun deleteById(id: LocalDate): Informe? {
        logger.debug { "InformeRepositoryMap ->\tdeleteById" }
        return informes.remove(id)
    }

    override fun update(element: Informe): Informe? {
        logger.debug { "InformeRepositoryMap ->\tupdate" }
        return updateById(element.day, element)
    }

    override fun updateById(id: LocalDate, element: Informe): Informe? {
        logger.debug { "InformeRepositoryMap ->\tupdateById" }
        if (getById(id) == null) return null
        informes[id] = element
        downgrade()
        return element
    }

    override fun upgrade(): List<Informe> {
        logger.debug { "InformeRepositoryMap ->\tupgrade" }
        informes.clear()
        val load = storageService.loadAll()
        saveAll(load, false)
        return load
    }

    override fun downgrade(): List<Informe> {
        logger.debug { "InformeRepositoryMap ->\tdowngrade" }
        return storageService.saveAll(informes.values.toList())
    }
}